File: backend/app/schemas/pet.py
"""Pet Pydantic schemas."""

from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


# Enums
PetSpecies = Literal["dog", "cat", "bird", "rabbit", "other"]
PetSex = Literal["male", "female", "unknown"]
PetStatus = Literal["available_for_adoption", "owned", "inactive"]


# Base schemas
class PetBase(BaseModel):
    """Base pet schema with common fields."""

    name: str = Field(..., min_length=1, max_length=100)
    species: PetSpecies
    breed: Optional[str] = Field(None, max_length=100)
    sex: PetSex
    birth_date: Optional[date] = None
    weight: Optional[float] = Field(None, gt=0, le=999.99)
    sterilized: bool = False


class PetCreate(PetBase):
    """Schema for creating a pet."""

    photo_key: Optional[str] = None
    owner_id: Optional[int] = None
    status: Optional[PetStatus] = "owned"


class PetUpdate(BaseModel):
    """Schema for updating a pet."""

    name: Optional[str] = Field(None, min_length=1, max_length=100)
    breed: Optional[str] = Field(None, max_length=100)
    sex: Optional[PetSex] = None
    birth_date: Optional[date] = None
    weight: Optional[float] = Field(None, gt=0, le=999.99)
    sterilized: Optional[bool] = None
    photo_key: Optional[str] = None
    status: Optional[PetStatus] = None


class PetResponse(PetBase):
    """Schema for pet response."""

    id: int
    tenant_id: int
    owner_id: Optional[int] = None
    status: PetStatus
    photo_url: Optional[str] = None
    photo_key: Optional[str] = None
    adopted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    created_by: int
    updated_by: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class PetDetailResponse(PetResponse):
    """Extended pet response with owner information."""

    owner: Optional[dict] = None  # Contains id, name, email if owner exists


class PaginatedPetResponse(BaseModel):
    """Schema for paginated pet list."""

    items: list[PetResponse]
    total: int
    page: int
    limit: int


# Upload-related schemas
class PresignedUploadRequest(BaseModel):
    """Request for presigned upload URL."""

    filename: str = Field(..., min_length=1, max_length=255)
    content_type: str = Field(..., min_length=1)
    file_size: int = Field(..., gt=0, le=5 * 1024 * 1024)  # 5MB max


class PresignedUploadResponse(BaseModel):
    """Response with presigned upload URL."""

    upload_url: str
    fields: dict
    photo_key: str
    cdn_url: str


# Filter schemas
class PetFilters(BaseModel):
    """Filters for pet listing."""

    owner_id: Optional[int] = None
    status: Optional[PetStatus] = None
    species: Optional[PetSpecies] = None
    search: Optional[str] = None
    page: int = Field(1, ge=1)
    limit: int = Field(20, ge=1, le=100)


# Adoption schemas
class AdoptPetRequest(BaseModel):
    """Request to assign owner (adopt) a pet."""

    owner_id: int = Field(..., gt=0)

File: frontend/src/types/pet.ts
/**
 * Pet-related TypeScript types and interfaces
 */

export type PetSpecies = 'dog' | 'cat' | 'bird' | 'rabbit' | 'other';
export type PetSex = 'male' | 'female' | 'unknown';
export type PetStatus = 'available_for_adoption' | 'owned' | 'inactive';

export interface Pet {
  id: number;
  tenant_id: number;
  name: string;
  species: PetSpecies;
  breed?: string | null;
  sex: PetSex;
  birth_date?: string | null; // ISO date string
  weight?: number | null; // kg
  sterilized: boolean;
  owner_id?: number | null;
  status: PetStatus;
  photo_url?: string | null;
  photo_key?: string | null;
  adopted_at?: string | null; // ISO datetime string
  created_at: string;
  updated_at: string;
  created_by: number;
  updated_by?: number | null;
}

export interface PetCreate {
  name: string;
  species: PetSpecies;
  breed?: string | null;
  sex: PetSex;
  birth_date?: string | null;
  weight?: number | null;
  sterilized?: boolean;
  photo_key?: string | null;
  owner_id?: number | null;
  status?: PetStatus;
}

export interface PetUpdate {
  name?: string;
  breed?: string | null;
  sex?: PetSex;
  birth_date?: string | null;
  weight?: number | null;
  sterilized?: boolean;
  photo_key?: string | null;
  status?: PetStatus;
}

export interface PaginatedPetResponse {
  items: Pet[];
  total: number;
  page: number;
  limit: number;
}

export interface PresignedUploadRequest {
  filename: string;
  content_type: string;
  file_size: number;
}

export interface PresignedUploadResponse {
  upload_url: string;
  fields: Record<string, string>;
  photo_key: string;
  cdn_url: string;
}

export interface PetFilters {
  owner_id?: number;
  status?: PetStatus;
  species?: PetSpecies;
  search?: string;
  page?: number;
  limit?: number;
}

export interface AdoptPetRequest {
  owner_id: number;
}

// Helper function to calculate pet age
export function calculateAge(birthDate: string | null | undefined): string {
  if (!birthDate) return 'Unknown';
  
  const birth = new Date(birthDate);
  const now = new Date();
  const years = now.getFullYear() - birth.getFullYear();
  const months = now.getMonth() - birth.getMonth();
  
  if (years === 0 && months === 0) {
    const days = Math.floor((now.getTime() - birth.getTime()) / (1000 * 60 * 60 * 24));
    return `${days} day${days !== 1 ? 's' : ''}`;
  }
  
  if (years === 0) {
    return `${months} month${months !== 1 ? 's' : ''}`;
  }
  
  if (months < 0) {
    return `${years - 1} year${years !== 2 ? 's' : ''}, ${12 + months} month${12 + months !== 1 ? 's' : ''}`;
  }
  
  if (months === 0) {
    return `${years} year${years !== 1 ? 's' : ''}`;
  }
  
  return `${years} year${years !== 1 ? 's' : ''}, ${months} month${months !== 1 ? 's' : ''}`;
}

// Helper to format species display name
export function formatSpecies(species: PetSpecies): string {
  return species.charAt(0).toUpperCase() + species.slice(1);
}

// Helper to format sex display name
export function formatSex(sex: PetSex): string {
  return sex.charAt(0).toUpperCase() + sex.slice(1);
}

// Helper to format status display name
export function formatStatus(status: PetStatus): string {
  return status
    .split('_')
    .map(word => word.charAt(0).toUpperCase() + word.slice(1))
    .join(' ');
}
